vm/src/vm.rs

```diff
@@ -944,14 +944,50 @@ impl VirtualMachine {
         }
     }
 
+    // Equivalent to check_class. Masks Attribute errors (into TypeErrors) and lets everything
+    // else go through.
+    fn _check_cls(&self, cls: &PyObjectRef, msg: String) -> PyResult {
+        self.get_attribute(cls.clone(), "__bases__").map_err(|e| {
+            // Only mask AttributeErrors.
+            if e.class().is(&self.ctx.exceptions.attribute_error) {
+                self.new_type_error(msg)
+            } else {
+                e
+            }
+        })
+    }
+
     pub fn _isinstance(&self, obj: &PyObjectRef, cls: &PyObjectRef) -> PyResult<bool> {
         if let Ok(typ) = PyTypeRef::try_from_object(self, cls.clone()) {
-            Ok(obj.isinstance(&typ))
+            if obj.class().issubclass(typ.clone()) {
+                Ok(true)
+            } else if let Ok(icls) =
+                PyTypeRef::try_from_object(self, self.get_attribute(obj.clone(), "__class__")?)
```

Comment:
is this line meaning `obj.class()` sometimes different from `obj.__class__`?

Reply:
Yup, check out the test cases for examples where this is manipulated.

Reply:
when `__class__` is replaced by python code, I think `obj.class()` need to be replaced rather than checked by here. I am good with this way for now, but could you add a FIXME comment here?

Reply:
Oh, sorry, I just saw the comment. Are you sure that's possible? I don't think `CPython` allows Python code modifying `tp_type` (and `obj.class()` grabs the equivalent of that if I'm not mistaken). It's the reason why on the next line we check:

```
if icls.is(&obj.class())
```

this way, if `__class__` is something different, we check if that is a subclass of `typ`.

Reply:
Thanks, I think that's possible but honestly I didn't checked how CPython handle it.